File: codility/NumberOfDiscIntersections.java
package codility;

import java.util.Arrays;

public class NumberOfDiscIntersections {

    public static void main(String[] args) {
        int[] arr = {1, 2147483647, 0};
        System.out.println(getNoOfIntersections(arr));
        System.out.println(2147483647 == Integer.MAX_VALUE);
    }

    public static int getNoOfIntersections(int[] arr) {
        int n = arr.length;
        long[] addArray = new long[n];
        long[] subArray = new long[n];
        int count = 0;
        int noOfIteration = 0;
        int noOfIntersection = 0;

        for(int i = 0; i < n; i++) {
            int b = i + arr[i];
            if(b>Integer.MAX_VALUE)
                b = Integer.MAX_VALUE;
            addArray[i] = b;
            subArray[i] = i - arr[i];
        }
        Arrays.sort(addArray);
        Arrays.sort(subArray);

        for(int j = 0; j < n; j++) {
            if(addArray[0] >= subArray[j]) {
                count += 1;
            }
            else {
                break;
            }
        }

        for(int k = 0; k < n; k++) {
            for(int s = count; s < n; s++) {
                if(addArray[k] >= subArray[s]) {
                    count += 1;
                }
                else {
                    break;
                }
            }
            noOfIntersection += count - 1 - noOfIteration;
            noOfIteration += 1;
        }

        if(noOfIntersection > 10000000) {
            return -1;
        }
        return noOfIntersection;
    }
}

File: codility/number_of_disc_intersections.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar  9 00:27:06 2018

@author: daniels
"""

def number_of_disc_intersection(arr):
    n = len(arr)
    add_array = [0] * n
    sub_array = [0] * n
    count = 0
    no_of_iterations = 0
    no_of_intersections = 0
    for i in range(0, n):
        add_array[i] = i + arr[i]
        sub_array[i] = i - arr[i]
    add_array.sort()
    sub_array.sort()
    
    for r in range(0,n):
        if(add_array[0] >= sub_array[r]):
            count += 1
        else:
            break
    for j in range(0, n):
        for k in range(count,n):
            if add_array[j] >= sub_array[k]:
                count += 1 
            else:
                break
        no_of_intersections += count - 1 - no_of_iterations
        no_of_iterations += 1    
    if (no_of_intersections > 10000000):
        return -1
    return no_of_intersections
